ci: Add Python 3.14-dev to the test matrix

jkseppan opened the pull request:
With continue-on-error set so failures don't break the build

<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary
<!-- Please describe the pull request, using the questions below as guidance, and link to any relevant issues and PRs:

- Why is this change necessary?
- What problem does it solve?
- What is the reasoning for this implementation?

Additionally, please summarize the changes in the title, for example "Raise ValueError on
non-numeric input to set_xlim" and avoid non-descriptive titles such as "Addresses
issue #8576".

If possible, please provide a minimum self-contained example.  If you have used
generative AI as an aid in preparing this PR, see

https://matplotlib.org/devdocs/devel/contribute.html#restrictions-on-generative-ai-usage
-->


## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

jkseppan committed:
Add Python 3.14-dev to the test matrix

With continue-on-error set so failures don't break the build

Co-authored-by: Oscar Gustafsson <oscar.gustafsson@gmail.com>

oscargus commented on .github/workflows/tests.yml:
```diff
@@ -88,6 +89,14 @@ jobs:
             pygobject-ver: '<3.52.0'
+          - name-suffix: "(Python 3.14 pre-release)"
+            os: ubuntu-24.04
+            python-version: '3.14.0-beta.3'
```
```suggestion
            python-version: '3.14.0-dev'
```
This will pick up the latest version (currently 3.14.0-beta-3 indeed), but avoid having to change it (until the first stable version is released).

oscargus commented:
Hmm... I'll check exactly what I do...

QuLogic commented on .github/workflows/tests.yml:
```diff
@@ -88,6 +89,14 @@ jobs:
+          - name-suffix: "(Python 3.14 pre-release)"
+            os: ubuntu-24.04
+            python-version: '3.14-dev'
+            # no pillow wheel yet, need extra packages to build from source
```
You should pull these from nightlies, just as in #30195, not build from source.

QuLogic replied:
Or see #28689.

QuLogic commented on .github/workflows/tests.yml:
```diff
@@ -88,6 +89,9 @@ jobs:
             python-version: '3.13'
             # https://github.com/matplotlib/matplotlib/issues/29732
             pygobject-ver: '<3.52.0'
+          - name-suffix: "(Python 3.14 pre-release)"
```
No need for `Python 3.14` in the suffix; it's already there at the start of the name.

QuLogic commented:
I suspect the timeouts are from GUI tests; for now, it may be a good idea to skip installing those, as is done for Python 3.13t.

jkseppan committed:
Skip interactive backends in 3.14 tests

The dependency wheels are not available for 3.14, and
the tkagg tests fail by running into timeouts. Use the
same if statement already used for 3.13t.

Also skip interactive-backend tests on the 3.14 pre-release.

jkseppan commented:
Now it doesn't hit the build timeout. The determinism tests fail by timing out, so there may be something in how we start subprocesses that's not working.

tacaswell commented:
The tests do pass for me on a locally built 3.14 so I do not think it is a generic thing we are doing wrong but rather an odd interaction with CI machines.

QuLogic commented:
3.14 tests were added in #30918.
